File: hardening/case_fold_oracle.rb
# Independent-runtime oracle: prints the Unicode full case fold of one string, then its lowercase.
# The argument bytes are tagged as UTF-8 and the output stream is set to UTF-8 so the result does not
# depend on the locale the reader happens to run under.
# Usage: ruby case_fold_oracle.rb '<text>'
STDOUT.set_encoding('UTF-8')
text = ARGV[0].dup.force_encoding('UTF-8')
puts text.downcase(:fold)
puts text.downcase

File: hardening/case_fold_oracle.php
<?php
// Independent-runtime oracle: prints the Unicode full case fold of one string, then its lowercase.
// Usage: php case_fold_oracle.php '<text>'
echo mb_convert_case($argv[1], MB_CASE_FOLD, 'UTF-8'), "\n";
echo mb_strtolower($argv[1], 'UTF-8'), "\n";
